From 2af74b40135aaa04d3ddcf2a8d110a9f6f392368 Mon Sep 17 00:00:00 2001
From: Georg Ringer <georg.ringer@gmail.com>
Date: Thu, 22 Nov 2018 08:46:21 +0100
Subject: [PATCH] [FEATURE] Make SiteLanguage available in TypoScript

The current language configuration should be available in TS as well.

Resolves: #86973
Releases: master
Change-Id: Iae71b550b3d000b12b4e127fee751740c3a36bf7
---

diff --git a/typo3/sysext/core/Documentation/Changelog/9.5.x/Feature-86973-TypoScriptGetTextPropertySiteLanguage.rst b/typo3/sysext/core/Documentation/Changelog/9.5.x/Feature-86973-TypoScriptGetTextPropertySiteLanguage.rst
new file mode 100644
index 0000000..6366893
--- /dev/null
+++ b/typo3/sysext/core/Documentation/Changelog/9.5.x/Feature-86973-TypoScriptGetTextPropertySiteLanguage.rst
@@ -0,0 +1,33 @@
+.. include:: ../../Includes.txt
+
+==========================================================
+Feature: #86973 - TypoScript getText property siteLanguage
+==========================================================
+
+See :issue:`86973`
+
+Description
+===========
+
+Site language configuration can now be accessed via the :typoscript:`getText` property `siteLanguage`
+in TypoScript.
+
+Examples:
+
+.. code-block:: typoscript
+
+	page.10 = TEXT
+	page.10.data = siteLanguage:navigationTitle
+	page.10.wrap = This is the title of the current site language: |
+
+.. code-block:: typoscript
+
+	page.10 = TEXT
+	page.10.dataWrap = The current site language direction is {siteLanguage:direction}
+
+Impact
+======
+
+Accessing the current site language configuration is now possible in TypoScript.
+
+.. index:: PHP-API, ext:frontend, NotScanned
diff --git a/typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php b/typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php
index 3b1342e..fc01d90 100644
--- a/typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php
+++ b/typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php
@@ -48,6 +48,7 @@
 use TYPO3\CMS\Core\Service\FlexFormService;
 use TYPO3\CMS\Core\Service\MarkerBasedTemplateService;
 use TYPO3\CMS\Core\Site\Entity\Site;
+use TYPO3\CMS\Core\Site\Entity\SiteLanguage;
 use TYPO3\CMS\Core\TimeTracker\TimeTracker;
 use TYPO3\CMS\Core\TypoScript\Parser\TypoScriptParser;
 use TYPO3\CMS\Core\TypoScript\TypoScriptService;
@@ -5100,6 +5101,16 @@
                             $retVal = ArrayUtility::getValueByPath($site->getConfiguration(), $key, '.');
                         }
                         break;
+                    case 'sitelanguage':
+                        $request = $GLOBALS['TYPO3_REQUEST'] ?? null;
+                        $siteLanguage = $request ? $request->getAttribute('language') : null;
+                        if ($siteLanguage instanceof SiteLanguage) {
+                            $config = $siteLanguage->toArray();
+                            if (isset($config[$key])) {
+                                $retVal = $config[$key];
+                            }
+                        }
+                        break;
                 }
             }
 
diff --git a/typo3/sysext/frontend/Tests/Unit/ContentObject/ContentObjectRendererTest.php b/typo3/sysext/frontend/Tests/Unit/ContentObject/ContentObjectRendererTest.php
index dc01283..d1294a9 100644
--- a/typo3/sysext/frontend/Tests/Unit/ContentObject/ContentObjectRendererTest.php
+++ b/typo3/sysext/frontend/Tests/Unit/ContentObject/ContentObjectRendererTest.php
@@ -25,6 +25,7 @@
 use TYPO3\CMS\Core\Context\UserAspect;
 use TYPO3\CMS\Core\Context\WorkspaceAspect;
 use TYPO3\CMS\Core\Core\ApplicationContext;
+use TYPO3\CMS\Core\Http\Uri;
 use TYPO3\CMS\Core\LinkHandling\LinkService;
 use TYPO3\CMS\Core\Log\Logger;
 use TYPO3\CMS\Core\Package\PackageManager;
@@ -33,6 +34,7 @@
 use TYPO3\CMS\Core\Resource\ResourceFactory;
 use TYPO3\CMS\Core\Resource\ResourceStorage;
 use TYPO3\CMS\Core\Site\Entity\Site;
+use TYPO3\CMS\Core\Site\Entity\SiteLanguage;
 use TYPO3\CMS\Core\TimeTracker\TimeTracker;
 use TYPO3\CMS\Core\TypoScript\TemplateService;
 use TYPO3\CMS\Core\Utility\DebugUtility;
@@ -1692,7 +1694,7 @@
     }
 
     /**
-     * Checks if getData() works with type "context"
+     * Checks if getData() works with type "site"
      *
      * @test
      */
@@ -1714,6 +1716,23 @@
     }
 
     /**
+     * Checks if getData() works with type "siteLanguage"
+     *
+     * @test
+     */
+    public function getDataWithTypeSiteLanguage(): void
+    {
+        $site = new SiteLanguage(1, 'de-de', new Uri('/'), [
+            'title' => 'languageTitle',
+            'navigationTitle' => 'German'
+        ]);
+        $serverRequest = $this->prophesize(ServerRequestInterface::class);
+        $serverRequest->getAttribute('language')->willReturn($site);
+        $GLOBALS['TYPO3_REQUEST'] = $serverRequest->reveal();
+        $this->assertEquals('German', $this->subject->getData('siteLanguage:navigationTitle'));
+    }
+
+    /**
      * Checks if getData() works with type "parentRecordNumber"
      *
      * @test
